**IMPORTS AND DATA PREPROCESSING**

In [1]:
import os
import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow import keras
from keras.datasets import imdb
from keras.preprocessing import sequence

2024-03-09 16:43:48.090267: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-03-09 16:43:48.090315: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-03-09 16:43:48.092014: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-03-09 16:43:48.101823: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-03-09 16:43:49.945426: W tensorflow/compiler/tf2

In [2]:
(X_train,y_train), (X_test, y_test) = imdb.load_data(num_words = 88584)

In [3]:
X_train[15]

[1,
 6,
 964,
 437,
 7,
 58,
 43,
 1402,
 11,
 6,
 171,
 3978,
 701,
 570,
 997,
 4179,
 1880,
 6,
 5211,
 5288,
 550,
 1270,
 1222,
 4891,
 519,
 6141,
 5,
 236,
 21,
 24,
 222,
 6,
 3934,
 31726,
 40,
 109,
 19,
 1513,
 8,
 1513,
 893,
 8016,
 5,
 1493,
 11,
 6,
 20619,
 5,
 25,
 80,
 28,
 14,
 806,
 1338,
 18,
 6,
 20,
 13,
 66,
 188,
 6,
 462,
 46,
 7,
 4,
 884,
 445,
 40,
 2699,
 32766,
 9671,
 4,
 114,
 16,
 1198,
 200,
 5400,
 5,
 701,
 2975,
 89,
 111,
 22416,
 26,
 1688,
 8,
 97,
 6,
 2476,
 1431,
 2863,
 2134,
 23,
 6,
 519,
 1522,
 1527,
 470,
 21,
 161,
 36,
 413,
 22416,
 143,
 49693,
 4756,
 14,
 16,
 398,
 18,
 15859,
 6286,
 12,
 299,
 40,
 6,
 374,
 500,
 13,
 122,
 358,
 4,
 68598,
 1530,
 5128,
 11,
 4,
 2092,
 151]

In [4]:
len(X_train[15])

129

In [5]:
len(X_train[8])

233

In [6]:
y_train

array([1, 0, 0, ..., 0, 1, 0])

In [7]:
X_train = sequence.pad_sequences(X_train, 250)
X_test = sequence.pad_sequences(X_test, 250)

In [8]:
X_train[9]

array([    0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     1,    14,    20,    47,   111,   439,
        3445,    19,

**BUILDING THE MODEL**

In [9]:
model = keras.Sequential([
    keras.layers.Embedding(88584, 128),
    keras.layers.LSTM(128),
    keras.layers.Dense(1, activation='sigmoid')
])

2024-03-09 16:43:58.973292: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:274] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected


In [10]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, None, 128)         11338752  
                                                                 
 lstm (LSTM)                 (None, 128)               131584    
                                                                 
 dense (Dense)               (None, 1)                 129       
                                                                 
Total params: 11470465 (43.76 MB)
Trainable params: 11470465 (43.76 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


**TRAINING THE MODEL**

In [11]:
model.compile(loss= "binary_crossentropy", optimizer= "rmsprop", metrics= ['acc'])
model.fit(X_train, y_train, epochs= 10)
keras.backend.clear_session()

Epoch 1/10
782/782 [==============================] - 302s 383ms/step - loss: 0.4530 - acc: 0.7890
Epoch 2/10
782/782 [==============================] - 399s 510ms/step - loss: 0.2905 - acc: 0.8862
Epoch 3/10
782/782 [==============================] - 347s 443ms/step - loss: 0.2314 - acc: 0.9136
Epoch 4/10
782/782 [==============================] - 272s 348ms/step - loss: 0.1847 - acc: 0.9322
Epoch 5/10
782/782 [==============================] - 257s 329ms/step - loss: 0.1442 - acc: 0.9492
Epoch 6/10
782/782 [==============================] - 275s 351ms/step - loss: 0.1145 - acc: 0.9613
Epoch 7/10
782/782 [==============================] - 246s 315ms/step - loss: 0.0915 - acc: 0.9701
Epoch 8/10
782/782 [==============================] - 250s 319ms/step - loss: 0.0669 - acc: 0.9787
Epoch 9/10
782/782 [==============================] - 259s 332ms/step - loss: 0.0523 - acc: 0.9833
Epoch 10/10
782/782 [==============================] - 248s 317ms/step - loss: 0.0422 - acc: 0.9866


**EVALUATING THE MODEL**

In [12]:
model.evaluate(X_test, y_test)

782/782 [==============================] - 67s 86ms/step - loss: 0.5387 - acc: 0.8700


[0.5387425422668457, 0.8700399994850159]

**The accuracy on the testing dataset is 87%, which pretty good!**